adagrams/game.py

```diff
@@ -1,11 +1,150 @@
+#     letter = random.choice(list(letter_pool_keys))
+#     if letter_pool[letter] > 0:
+#         drawn_letters.append(letter)
+
+#         #adding random letter to the list and decreasing the frequency of the letter
+#         letter_pool[letter]-= 1
+#    return drawn_letter
+
+
+   total_letters = sum(letter_pool.values())
+
+   while len(drawn_letters) < 10:
+        # Generating a random number between 1 and the total number of letters
+        rand_num = random.randint(1, total_letters)
+
+        # Finding  the corresponding letter
+        cumulative_count = 0
+        for letter, frequency in letter_pool.items():
+            cumulative_count += frequency
+            if rand_num <= cumulative_count:
+                drawn_letters.append(letter)
+                letter_pool[letter] -= 1
+                total_letters -= 1
+                break
+
+
+   return drawn_letters
+
 
 def uses_available_letters(word, letter_bank):
-    pass
 
+#    converting all the letters in letter bank to upper case
+    letter_bank = [letter.upper() for letter in letter_bank]
+# making a dictionary for a letter_bank to check the frequency of letters
+    lookup = {}
+    for letter in letter_bank:
+        if letter in lookup:
+            lookup[letter] += 1
+        else:
+            lookup[letter] = 1
+    #converting all the letters of word into uppercase and checking the frequebcy of the letter, if found decrementing the value or else returning false
+
+    for letter in word.upper():
+        if letter not in lookup or lookup[letter] == 0:
+            return False
+        else:
+            lookup[letter] -= 1
+    return True
+
+#Approaching as per draw_letters function with nested loops
+
+# def score_word(word):
+# #    converting all the letters in letter bank to upper case
+#     word = [letter.upper() for letter in word]
+#     total_score = 0
+#     # making a dictionary with letters and it's score value
+#     score_value = {
+#         'AEIOULNRST': 1,'DG': 2, 'BCMP': 3, 'FHVWY': 4, 'K': 5, 'JX': 8, 'QZ': 10
+#         }
+
+#     for letter in word:
+#         for scoring_letter_group, score in score_value.items():
+#             if letter in scoring_letter_group:
+#                 total_score+= score
+#     if len(word) > 6 and len(word) <=10:
+#         total_score += 8
+#     return total_score
+
+
+
+# optimal solution where dictionary is created for individual letter and score value without nested loops
 def score_word(word):
-    pass
+    score_value = {
+        'A': 1, 'E': 1, 'I': 1, 'O': 1, 'U': 1, 'L': 1, 'N': 1, 'R': 1, 'S': 1, 'T': 1,
+        'D': 2, 'G': 2,
+        'B': 3, 'C': 3, 'M': 3, 'P': 3,
+        'F': 4, 'H': 4, 'V': 4, 'W': 4, 'Y': 4,
+        'K': 5,
+        'J': 8, 'X': 8,
+        'Q': 10, 'Z': 10
+    }
+    word = word.upper()
+    total_score = 0
+
+    total_score+= sum(score_value.get(letter, 0) for letter in word )
+
+    if len(word) > 6 and len(word) <=10:
+        total_score += 8
```

Comment:
Using constant variables will make your code more self-documenting and easier to understand. If I don't know anything about this game, I'm left wondering what the significance of the integer literals. are. 

Consider something like:
```python
BONUS_LENGTH_MIN = 7
BONUS_LENGTH_MAX = 10
BONUS_POINTS = 8

if BONUS_LENGTH_MIN <= len(word) <= BONUS_LENGTH_MAX:
        total_score += BONUS_POINTS
```